nanopass/manual.rst

```diff
@@ -0,0 +1,213 @@
+
+Nanopass Framework
+==================
+
+This document describes how to use the nanopass framework, what is allowed, and
+what not. It only describes the framework's interface, not how things are
+implemented internally.
+
+Nanopass Compiler Overview
+--------------------------
+
+A nanopass compiler is a compiler implemented as a series of narrow *passes*,
+which *transform*, *analyze*, or otherwise produce abstract syntax trees (=AST)
+according to well-defined grammars.
+
+The idea is that passes focus on small, specific tasks, with tree traversal
+boilerplate generated automatically.
+
+Concepts
+--------
+
+* a *language* (in the context of the nanopass framework) is a formal grammar.
+* a *terminal* is ...
+* a *form* is a named schema for a term, made up of zero or more sub-terms
+* a *non-terminal* is ...
+* a *meta-variable* is a name ranging over a terminal or non-terminal. It may be viewed as an alias.
+
+.. TODO rewrite this section. It should use the correct terminology while still
+        being approachable to someone not overly familiar with formal grammars
+
+Usage
+-----
+
+Language Definition
+~~~~~~~~~~~~~~~~~~~
+
+Languages are defined via the `defineLanguage` macro, which has two forms.
+
+From Scratch
+~~~~~~~~~~~~
+
+The first form `defineLanguage` form takes two arguments, an identifier and
+a body, and defines a language from scratch. If the body is valid, a NimSkull
+type is bound to the identifier.
+
+.. note::
+
+  The type is not meant to be used as a type for values. Rather, it
+  acts as a namespace through which entities defined in the language are
+  accessed.
+
+The body consists of a sequence of terminal and non-terminal definitions.
+
+.. code-block:: nim
+    :test: "nim c $1"
+
+  import nanopass/nanopass
+
+  defineLanguage L0:
+    i(int)         # definition of a terminal
+    expr(e) ::= i  # definition of a non-terminal
+
+  # this defines a language `L0` with:
+  # * a single terminal of type `int`, ranged over by meta-variable `i`
+  # * a non-terminal with name `expr`, ranged over by meta-variable `e`. Where
+  #   this non-terminal is expected, an `int` value is allowed
+
+Meta-variables must be unique.
+
+.. code-block:: nim
+    :test: "nim c $1"
+    :status: 1
+
+  import nanopass/nanopass
+
+  defineLanguage L0:
+    i(int)
+    i(float) # error: 'i' name already in use
+
+.. code-block:: nim
+    :test: "nim c $1"
+    :status: 1
+
+  import nanopass/nanopass
+
+  defineLanguage L0:
+    i(int)
+    integer(i) ::= i # error: 'i' name already in use
```

Comment:
Doesn't have to change, but this feels like the error message should instead be more like "attempting to range over a metavariable, provide a type instead", unless I'm forgetting something (highly likely). I get the names exist in the same space, but I thought they were still distinct in terms of type (ranging over something metavar vs type name).

I'll be doing some rereading to remind myself of these bits, so I might update the comment in a few days.

Reply:
ok, I see where I'm confused and how things went off the rails from there, closing this comment.